File: LinkedList/cpp/LinkedList.cpp
#include <iostream>
#include "LinkedList.h"

template <typename T>
LinkedList<T>::LinkedList()
    : sizeOfList{0}, root{nullptr} {
}

template <typename T>
LinkedList<T>::LinkedList(T data){
    this->sizeOfList = 1;

    Node<T> *newNode = new Node<T>(data);
    this->root = newNode;

    delete newNode;
}


template <typename T>
LinkedList<T>::~LinkedList(){
}

template <typename T>
void LinkedList<T>::insert(T data){
    ++this->sizeOfList;
    Node<T> *newNode = new Node<T>(data);

    if(this->root == nullptr){
        this->root = newNode;    
    }
    else{
        newNode->setNextNode(this->root);
        this->root = newNode;
    }
}

template <typename T>
void LinkedList<T>::remove(T data){
    if(this->root == nullptr){
        return;
    }

    --this->sizeOfList;
    if(this->root->getData() == data){
        this->root = this->root->getNextNode();
    }
    else{
        remove(data, this->root, this->root->getNextNode());
    }
}

template <typename T>
void LinkedList<T>::remove(T data, Node<T> *previousNode, Node<T> *actualNode){
    if(actualNode == nullptr){
        return;
    }
    if(actualNode->getData() == data){
        previousNode->setNextNode(actualNode->getNextNode());
        delete actualNode;
    }
    else{
        remove(data, actualNode, actualNode->getNextNode());
    }
}

template <typename T>
void LinkedList<T>::traverseList(){
    traverseList(this->root);
}

template <typename T>
void LinkedList<T>::traverseList(Node<T> *node){
    if(node == nullptr){
        std::cout << "NULL";
        return;
    }
    std::cout << node->getData() << " -> ";
    traverseList(node->getNextNode());
}

template <typename T>
int LinkedList<T>::size(){
    return this->sizeOfList;
}

template <typename T>
bool LinkedList<T>::empty(){
    if(this->root == nullptr){
        return true;
    }
    return false;
}

template <typename T>
T LinkedList<T>::middle(){
    Node<T> *faster = root;
    Node<T> *slower = root;

    while(faster->getNextNode() != nullptr && faster->getNextNode()->getNextNode() != nullptr){
        faster = faster->getNextNode()->getNextNode();
        slower = slower->getNextNode();
    }

    return slower->getData();
}

template <typename T>
void LinkedList<T>::reverse(){
    Node<T> *nextNode = nullptr;
    Node<T> *previousNode = nullptr;
    Node<T> *currentNode = this->root;

    while(currentNode != nullptr){
        nextNode = currentNode->getNextNode();
        currentNode->setNextNode(previousNode);

        previousNode = currentNode;
        currentNode = nextNode;
    }

    this->root = previousNode;
}

File: LinkedList/python/linked_list.py
from node import Node

class LinkedList:
    
    _size_of_list = 0

    def __init__(self, data=None):
        self._root = Node()
        
        if data != None:
            self.insert(data)
    
    def __len__(self):
        return self._size_of_list

    @property
    def root(self):
        return self._root
    
    @root.setter
    def root(self, node):
        self._root = node

    def insert(self, data):
        self._size_of_list += 1

        new_node = Node(data)
        new_node.next_node = self.root
        self.root = new_node

    def remove(self, data):
        if self.root.data == None:
            return
        
        self._size_of_list -= 1
        if self.root.data == data:
            self.root = self.root.next_node
        else:
            LinkedList.__remove(data, self.root, self.root.next_node)

    @staticmethod
    def __remove(data, previous_node, actual_node):
        if actual_node.data == None:
            return
        if actual_node.data == data:
            previous_node.next_node = actual_node.next_node
        else:
            LinkedList.__remove(data, actual_node, actual_node.next_node)

    def traverse_list(self):
        it = self.root

        while it.next_node != None:
            print(it.data, end=' -> ')
            it = it.next_node
        print(it.next_node, end = '')

    def empty(self):
        if self.root.data == None:
            return True
        return False

    def middle(self):
        faster = self.root
        slower = self.root

        while faster.next_node.data != None and faster.next_node.next_node.data != None:
            faster = faster.next_node.next_node
            slower = slower.next_node

        return slower.data

    def reverse(self):
        previous_node = Node()
        after_node = Node()

        current_node = Node()
        current_node = self.root

        while current_node.data != None:
            after_node = current_node.next_node
            current_node.next_node = previous_node

            previous_node = current_node
            current_node = after_node

        self.root = previous_node

